source/postprocess/visualization.cc

```diff
@@ -548,28 +620,84 @@ namespace aspect
-                                    std::set<std::string> &output_data_names_set)
+      void
+      track_output_field_names_and_units(const std::vector<std::string> &output_data_names,
+                                         const std::vector<std::string> &output_units,
+                                         std::map<std::string,std::string> &output_data_names_and_units)
       {
-        // Copy the vector elements into a std::set. This discards
-        // duplicates as mentioned in the comment above.
-        const std::set<std::string> set_of_names(output_data_names.begin(),
-                                                 output_data_names.end());
+        AssertDimension (output_data_names.size(), output_units.size());
+
+        // First check that none of the data names are already in the list:
+        for (const auto &name : output_data_names)
+          AssertThrow(output_data_names_and_units.find(name) == output_data_names_and_units.end(),
+                      ExcMessage("The output variable <" + name + "> already exists in the list of output "
+                                 "variables. Make sure there is no duplication in the names of visualization output "
+                                 "variables, otherwise output files may be corrupted."));
+
+        // Then insert the new names into the map. It is possible that
+        // some names appear multiple times (for vector-valued output)
+        // and in that case we need to make sure that the provided
+        // units are identical
+        for (unsigned int i=0; i<output_data_names.size(); ++i)
+          {
+            AssertThrow(
+              // either the name has not been added before...
+              (output_data_names_and_units.find(output_data_names[i])
+               == output_data_names_and_units.end())
+              ||
+              // or if it has, then the unit then added must match
+              // what we're about to add here
+              (output_data_names_and_units[output_data_names[i]]
+               == output_units[i]),
+              ExcMessage("The output variable <" + output_data_names[i]
+                         + "> had previously been added with physical units <"
+                         + output_data_names_and_units[output_data_names[i]]
+                         + "> but is now added again with units <"
+                         + output_units[i] + ">."));
+
+            // Now add the name/units. Don't worry about whether it previously
+            // had already been in the map -- if so, the assertion makes
+            // sure that we overwrite with the same value.
+            output_data_names_and_units[output_data_names[i]] = output_units[i];
+          }
+      }
+
 
-        for (const auto &name: set_of_names)
+
+      /**
+       * Keep track of the names and physical units of output quantities. The
+       * of entries in the two input arguments must be the same.
```

Comment:
This sentence is missing some
:slightly_smiling_face:

Reply:
I should have been more clear and less witty in my comment, the second sentence is not clear and is likely missing a word. Which entries in which input arguments need to be the same (there are three)?